File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int l = matrix.length - 1;
        int h = matrix[0].length - 1;
        int l1 = 0;
        int l2 = matrix.length - 1;
        int h1 = 0;
        int h2 = matrix[0].length - 1;
        while (l1 <= l2 && h1 <= h2)
        {
            int midl = (l2 - l1)/2 + l1;
            int midh = (h2 - h1)/2 + h1;
            //System.out.println(matrix[midl][midh]);
            if (matrix[midl][midh] == target)
            {
                return true;
            }
            else if (matrix[midl][midh] < target)
            {
                if (matrix[midl][h] == target)
                {
                    return true;
                }
                else if (matrix[midl][h] > target)
                {
                    l1 = midl;
                    h1 = midh + 1;
                    l2 = midl;
                }
                else
                {
                    l1 = midl + 1;
                    h1 = 0;
                }
            }
            else
            {
                if (matrix[midl][0] == target)
                {
                    return true;
                }
                else if (matrix[midl][0] > target)
                {
                    l2 = midl - 1;
                    h2 = h;
                }
                else
                {
                    l2 = midl;
                    h2 = midh - 1;
                    l1 = midl;
                }
            }
        }
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROW = len(matrix) - 1
        COL = len(matrix[0]) - 1 
        lr, lc = 0, 0
        hr, hc = ROW, COL
        while lr <= hr and lc <= hc:
            midr = (lr + hr) // 2
            midc = (lc + hc) // 2
            if matrix[midr][midc] == target:
                return True
            elif target < matrix[midr][midc]:
                if target < matrix[midr][lc]:
                    hr = midr - 1
                else:
                    lr = midr
                    hr = midr
                    hc = midc - 1
            else:
                if target > matrix[midr][hc]:
                    lr = midr + 1
                else:
                    hr = midr
                    lc = midr
                    lc = midc + 1
        return False

